tests/topotests/ospf_multi_vrf_bgp_route_leak/test_ospf_multi_vrf_bgp_route_leak.py

```diff
@@ -0,0 +1,247 @@
+
+    # Interconect router 1, 3
+    switch = tgen.add_switch("s1-3")
+    switch.add_link(tgen.gears["r1"])
+    switch.add_link(tgen.gears["r3"])
+
+    # Interconect router 1, 4
+    switch = tgen.add_switch("s1-4")
+    switch.add_link(tgen.gears["r1"])
+    switch.add_link(tgen.gears["r4"])
+
+
+def setup_module(mod):
+    logger.info("OSPF Multi VRF Topology with BGP route leaking:\n {}".format(TOPOLOGY))
+
+    tgen = Topogen(build_topo, mod.__name__)
+    tgen.start_topology()
+
+    r1_vrf_setup_cmds = [
+        "ip link add name neno type vrf table 11",
+        "ip link add name ray type vrf table 12",
+        "ip link set dev neno up",
+        "ip link set dev ray up",
+        "ip link set dev r1-eth2 vrf neno up",
+        "ip link set dev r1-eth3 vrf ray up",
+    ]
+
+    # Starting Routers
+    router_list = tgen.routers()
+
+    # Create VRFs on r2 and bind to interfaces
+    for cmd in r1_vrf_setup_cmds:
+        tgen.net["r1"].cmd(cmd)
+
+    logger.info("Testing OSPF VRF support")
+
+    for rname, router in router_list.items():
+        logger.info("Loading router %s" % rname)
+        router.load_config(
+            TopoRouter.RD_ZEBRA, os.path.join(CWD, "{}/zebra.conf".format(rname))
+        )
+        router.load_config(
+            TopoRouter.RD_OSPF, os.path.join(CWD, "{}/ospfd.conf".format(rname))
+        )
+        if rname in ["r1"]:
+            router.load_config(
+                TopoRouter.RD_BGP, os.path.join(CWD, "{}/bgpd.conf".format(rname))
+            )
+
+    # Initialize all routers.
+    tgen.start_router()
+    for router in router_list.values():
+        if router.has_version("<", "4.0"):
+            tgen.set_error("unsupported version")
+
+
+def teardown_module(mod):
+    "Teardown the pytest environment"
+    tgen = get_topogen()
+    tgen.stop_topology()
+
+
+# Shared test function to validate expected output.
+def compare_show_ip_route_vrf(rname, expected, vrf_name):
+    """
+    Calls 'show ip route vrf [vrf_name] route' and compare the obtained
+    result with the expected output.
+    """
+    tgen = get_topogen()
+    current = topotest.ip4_route_zebra(tgen.gears[rname], vrf_name)
+    ret = topotest.difflines(
+        current, expected, title1="Current output", title2="Expected output"
+    )
+    return ret
+
+
+def test_ospf_convergence():
+    "Test OSPF daemon convergence"
+    tgen = get_topogen()
+
+    if tgen.routers_have_failure():
+        pytest.skip("skipped because of router(s) failure")
+
+    for rname, router in tgen.routers().items():
+        logger.info('Waiting for router "%s" convergence', rname)
+
+        for vrf in ["default", "neno", "ray"]:
+            # Load expected results from the command
+            reffile = os.path.join(CWD, "{}/ospf-vrf-{}.txt".format(rname, vrf))
+            if os.path.exists(reffile):
+                expected = open(reffile).read()
+
+                # Run test function until we get an result. Wait at most 60 seconds.
```

Comment:
1. We are directly doing route convergence , better to check neighbour convergence and then go to route convergence .. 
2.Here we are waiting 80 secs.

Reply:
I thought if we get route convergence the we cover the case of neighbor convergence since we won't convergence on routes if we don't converge on neighbors first.

Reply:
2. Fixed, thanx!

Reply:
> 

Agree, but if scripts fails because of neighbour convergence, error will be misleading telling routes not converged.

We can assert  on neighbour convergence failure and exit, on success we can continue and  assert on route convergence failure and exit.  It helps in failure analysis.